diagnostics: shorten paths of unique symbols

da-x opened the pull request:
This is a step towards implementing a fix for #50310, and continuation of the discussion in [Pre-RFC: Nicer Types In Diagnostics - compiler - Rust Internals](https://internals.rust-lang.org/t/pre-rfc-nicer-types-in-diagnostics/11139). Impressed upon me from previous discussion in #21934 that an RFC for this is not needed, and I should just come up with code.

The recent improvements to `use` suggestions that I've contributed have given rise to this implementation. Contrary to previous suggestions, it's rather simple logic, and I believe it only reduces the amount of cognitive load that a developer would need when reading type errors.

-----

If a symbol name can only be imported from one place, and as long as it was not glob-imported anywhere in the current crate, we can trim its printed path to the last component.

This has wide implications on error messages with types, for example, shortening `std::vec::Vec` to just `Vec`, as long as there is no other `Vec` importable from anywhere.

rust-highfive commented:
r? @estebank

(rust_highfive has picked a reviewer for you, use r? to override)

estebank commented:
This is an ingenious solution that as shown can get us quite far and love it's simplicity. I'm concerned about two things:

* the memory implications of this change on larger projects, but the Symbols are interned so this shouldn't be a huge impact and is at worst linear

* now we use the name only, even if the type is not in scope

This last point has a couple of implications. Types that haven't been imported might produce errors that don't let you figure out how to suggest them. Also I didn't see the whole PR yet but there are `use` suggestions

da-x commented:
Thanks for looking into this! Several things -

If we limit this to prelude imports then `Vec` benefits but `std::collections::HashMap` doesn't, which IMHO does not look too well.

Adding scope checks to check, for example, whether `std::collections::HashMap` is imported at the error site has some drawbacks. 
1) I think there's some value in having consistency across all the type errors of a single crate build invocation, considering that we can have one module do `use std::collections::HashMap` and another not have the import statement. 
2) Whether `HashMap` import is actually used or not is not tied entirely to whether type errors would refer to `HashMap` at the scope in question, and I believe developers would like to avoiding unused imports just for the sake of shorter types being printed. 

As it is now in the PR, it is now does not affect `use` statements. If we short paths for non-prelude types like `HashMap` then newcomers might guess wrong that it's in the prelude, but that's okay because the compiler then just tells them to `use std::collections::HashMap` via the existing use suggestions. That's why the PR only does this for importable items - because the use suggestions backs them up. It's an extra step. Maybe it's not enough?

estebank commented:
The point about consistency doesn't sway me much as I would like people to be able to copy things from the output and paste them where the error happened and have things just work, but the later point makes a lot of sense: people will just get the follow up suggestion for imports. This last thing is something we already rely on other cases and I feel is a perfectly fine workflow. I'll review today but now I think this is a perfectly reasonable approach. It will still obviously not address the case of repeated identifiers but it's already a huge improvement.

lzutao commented on src/test/ui/associated-type-bounds/assoc-type-eq-with-dyn-atb-fail.stderr:
```diff
@@ -1,8 +1,8 @@
-error[E0277]: the trait bound `std::string::String: std::marker::Copy` is not satisfied
+error[E0277]: the trait bound `String: std::marker::Copy` is not satisfied
```
The trait bound here isn't shorten.

da-x replied:
I'll correct this. This is because I treated all the different kind of `Namespace` as a single one, there are some places in `std` where you have both a macro and a type with the same same. This will be fixed.

petrochenkov commented on src/librustc_resolve/lib.rs:
```diff
@@ -1440,6 +1451,49 @@ impl<'a> Resolver<'a> {
         }
     }
 
+    fn find_unique_symbols(&mut self) {
```
Could you move this to `diagnostics.rs`?

da-x replied:
Done

petrochenkov commented:
Could you also run perf on this before merging?

This PR doesn't do too much harm, so I guess it's ok for it to exist until something more principled is implemented.
(Didn't look at the details.)

da-x commented:
Going over the long list of test changes (~1250 files!), found some bits to work out:

* The `String`-related hints in `src/test/ui/span/issue-39018.stderr` are gone.
* `std::fmt::Display` doesn't get shortened. Is there another `Display` somewhere? :(
* `std::result::Result` doesn't get shortened. Colliding with `std::io::Result`? This being a sister type of `Option`, which does get handled, it's a bit unfortunate.
* There are still a few places where `std::marker::Copy` is not shortened without a clear reason why.

estebank commented:
> * The String-related hints in `src/test/ui/span/issue-39018.stderr` are gone.

`rustc_on_unimplemented` uses the string representation of the path for types to select what message to give. We'll need to change the code in `src/librustc_trait_selection/traits/error_reporting/mod.rs:report_selection_error` to always use the full path.

bors commented:
:umbrella: The latest upstream changes (presumably #74117) made this pull request unmergeable. Please resolve the merge conflicts.

da-x commented on src/librustc_middle/ty/context.rs:
```diff
@@ -935,6 +935,9 @@ pub struct GlobalCtxt<'tcx> {
     maybe_unused_trait_imports: FxHashSet<LocalDefId>,
     maybe_unused_extern_crates: Vec<(LocalDefId, Span)>,
+
+    pub unique_symbols: FxHashMap<DefId, Symbol>,
```
Done

da-x commented:
> `rustc_on_unimplemented` uses the string representation of the path for types to select what message to give. We'll need to change the code in `src/librustc_trait_selection/traits/error_reporting/mod.rs:report_selection_error` to always use the full path.

Thanks for the hint - I got it fixed.

da-x commented:
@estebank Separating uniqueness for trait symbols and type symbols seems to sort out the overlap between `std::fmt::Display` and `std::path::Display`.
Also, allowing type aliases to be superseded in uniqueness by the types of the same name will take care of the `std::io::Result` vs `std::result::Result` overlap, with `std::result::Result` winning over `Result`.

WDYT?

nikomatsakis commented:
I've not read the diff, but I'm very excited to see progress on this problem! 💖

estebank commented:
> @estebank Separating uniqueness for trait symbols and type symbols seems to sort out the overlap between `std::fmt::Display` and `std::path::Display`.
> Also, allowing type aliases to be superseded in uniqueness by the types of the same name will take care of the `std::io::Result` vs `std::result::Result` overlap, with `std::result::Result` winning for Result.
>
> WDYT?

I'm of diverging opinions. I was originally going to dismiss this alternative out of hand as I fear it would be confusing to do this deduplication on a per type basis. But then I thought a bit about it and 1) it makes absolute sense to ignore type aliases (at least for now) because we never refer to them in type errors and 2) we already give a short description of what we are talking about (`expected type 'Result', found trait 'Display'`), which would make that work out nicely. The only case where I am slightly concerned is when both "expected" and "found" clash: `expected type 'Display', found trait 'Display'` could be confusing in of itself. Could we have a way to handle that case specifically? I feel that we can do these things as a follow up PR, what do you think?

da-x commented:
Yes, we can solve for specific ambiguities with later PRs. 

I've pushed remaining fixes discussed for `Result` and `Copy`.

Going to check the performance of `find_unique_symbols`, now that `lookup_import_candidates` is done on the good path, and not only during `use` suggestions.

Also, should we allow users to disable this with a `-Z` in case issues arise, to ease debugging in nightly?

estebank commented:
> Also, should we allow users to disable this with a -Z in case issues arise, to ease debugging in nightly?

Yes.

da-x commented:
Performance -

For rustc invoked under perf with callgraphs (`perf record  -F 99 --call-graph=dwarf `), when building the main ripgrep crate,`find_unique_symbols` takes a total of 2.54% of execution time.

```
-    2.54%     0.00%  rustc  librustc_driver-3de5a075e28cbed1.so  [.] rustc_resolve::diagnostics::<impl rustc_resolve::Resolver>::find_unique_symbols                                                                        ◆
   - rustc_resolve::diagnostics::<impl rustc_resolve::Resolver>::find_unique_symbols                                                                                                                                         ▒
      - 2.50% rustc_resolve::diagnostics::<impl rustc_resolve::Resolver>::lookup_import_candidates (inlined)                                                                                                                 ▒
           rustc_resolve::diagnostics::<impl rustc_resolve::Resolver>::lookup_import_candidates_from_module                                                                                                                  ▒
         - rustc_resolve::ModuleData::for_each_child (inlined)                                                                                                                                                               ▒
            - 2.18% rustc_resolve::Resolver::resolutions                                                                                                                                                                     ▒
               - rustc_resolve::build_reduced_graph::<impl rustc_resolve::Resolver>::build_reduced_graph_external (inlined)                                                                                                  ▒
                  - 1.17% rustc_metadata::rmeta::decoder::cstore_impl::<impl rustc_metadata::creader::CStore>::item_children_untracked                                                                                       ▒
                       rustc_metadata::rmeta::decoder::<impl rustc_metadata::creader::CrateMetadataRef>::each_child_of_item (inlined)                                                                                        ▒
                    1.01% rustc_resolve::build_reduced_graph::BuildReducedGraphVisitor::build_reduced_graph_for_external_crate_res                                                                                           ▒
     0.12%     0.00%  rustc  librustc_driver-3de5a075e28cbed1.so  [.] rustc_resolve::diagnostics::_$LT$impl$u20$rustc_resolve..Resolver$GT$::find_unique_symbols::_$u7b$$u7b$closure$u7d$$u7d$::h583030594edeb63a (inlined) 
```
@petrochenkov

bors commented:
:umbrella: The latest upstream changes (presumably #74245) made this pull request unmergeable. Please resolve the merge conflicts.

eddyb commented on src/librustc_middle/ty/print/pretty.rs:
```diff
@@ -28,9 +28,24 @@ macro_rules! p {
     (@write($($data:expr),+)) => {
         write!(scoped_cx!(), $($data),+)?
     };
+    (@write_full_path($($data:expr),+)) => {
+        with_crate_prefix(|| {
+            write!(scoped_cx!(), $($data),+)
+        })?
+    };
     (@print($x:expr)) => {
         scoped_cx!() = $x.print(scoped_cx!())?
     };
+    (@$method:ident(cond_full_path: $cond:expr, $($arg:expr),*)) => {
+        let res = if $cond {
+            with_crate_prefix(|| {
+                scoped_cx!().$method($($arg),*)
+            })
+        } else {
+            scoped_cx!().$method($($arg),*)
+        };
+        scoped_cx!() = res?
+    };
```
I believe you can do these manually around `p!` calls and avoid changing the macro.

da-x replied:
I tried that approach first, trying to avoid changing this macro, however needing to use `with_crate_prefix(||` closure around `p!` tripped it badly. I don't recall exactly what was the error, but I think it's related to `scoped_cx!()` which makes assumptions that are not workable. I can try again and post about it.

eddyb replied:
Oh I think it's because of the closure nesting, that makes sense. You should be able to remove the users of this post-#74344, as per my other comments, thankfully.

eddyb commented on src/librustc_middle/ty/print/pretty.rs:
```diff
@@ -1317,6 +1346,13 @@ impl<F: fmt::Write> Printer<'tcx> for FmtPrinter<'_, 'tcx, F> {
         define_scoped_cx!(self);
 
         if substs.is_empty() {
+            if !SHOULD_PREFIX_WITH_CRATE.with(|flag| flag.get()) {
+                match self.try_print_unique_def(def_id)? {
+                    (cx, true) => return Ok(cx),
+                    (cx, false) => self = cx,
+                }
+            }
+
             match self.try_print_visible_def_path(def_id)? {
```
I wonder if it might be useful to move the logic into `try_print_visible_def_path`.

da-x replied:
I thought so to, but figured that the name of `try_print_visible_def_path` would need to change to convey its dual purpose, and the coupled name name `try_print_visible_def_path_recur` too, so I'm not sure about it. Given that `try_print_visible_def_path` has a single call site, the benefit feels diminished to me.

eddyb replied:
But "visible def path" is what you're doing, isn't it? Just not absolute. Maybe I'm overthinking things.

petrochenkov replied:
Seems ok as is, marking as resolved.

eddyb commented on src/librustc_middle/ty/print/pretty.rs:
```diff
@@ -243,6 +258,20 @@ pub trait PrettyPrinter<'tcx>:
+        match unique_symbols.get(&def_id) {
+            None => return Ok((self, false)),
+            Some(symbol) => {
+                write!(&mut self, "{}", symbol)?;
```
Is this used (as opposed to `write_str`) elsewhere? I forget.

da-x replied:
Yes, this can be changed to `write_str`. Done.

eddyb commented on src/librustc_trait_selection/traits/error_reporting/mod.rs:
```diff
@@ -283,7 +283,7 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                             .map(|s| &s == "?")
                             .unwrap_or(false);
-                        let is_from = format!("{}", trait_ref.print_only_trait_path())
+                        let is_from = format!("{}", trait_ref.print_only_trait_path_full())
                             .starts_with("std::convert::From<");
```
String manipulation is really hacky, and code like this should instead check `trait_ref.def_id` for a lang item (if one exists) or a diagnostic item otherwise.

da-x replied:
Yeah, I agree it's hacky. But I think it should be a separate PR to tackle it. This PR is already a big change on its own.

estebank replied:
This predates the `lang_items` being in the codebase. We should change it as part of a separate PR.

eddyb commented on src/librustc_trait_selection/traits/error_reporting/mod.rs:
```diff
@@ -316,10 +316,11 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
-                            format!("{}", trait_ref.print_only_trait_path())
+                            format!("{}", trait_ref.print_only_trait_path_full())
                                 .starts_with("std::convert::From<std::option::NoneError");
-                        let should_convert_result_to_option = format!("{}", trait_ref)
-                            .starts_with("<std::option::NoneError as std::convert::From<");
+                        let should_convert_result_to_option =
+                            format!("{}", trait_ref.print_full_path())
+                                .starts_with("<std::option::NoneError as std::convert::From<");
```
`NoneError` should be a diagnostic item, instead of string matching.

estebank replied:
Addressed in https://github.com/rust-lang/rust/pull/74344

da-x replied:
Thanks to #74344, changes to `src/librustc_trait_selection/traits/error_reporting/mod.rs` reverted in this PR.

eddyb commented on src/librustc_middle/ty/print/pretty.rs:
```diff
@@ -1874,14 +1910,40 @@ impl fmt::Debug for TraitRefPrintOnlyTraitPath<'tcx> {
 
 impl ty::TraitRef<'tcx> {
     pub fn print_only_trait_path(self) -> TraitRefPrintOnlyTraitPath<'tcx> {
-        TraitRefPrintOnlyTraitPath(self)
+        TraitRefPrintOnlyTraitPath(self, false)
+    }
+    pub fn print_only_trait_path_full(self) -> TraitRefPrintOnlyTraitPath<'tcx> {
+        TraitRefPrintOnlyTraitPath(self, true)
+    }
+}
+
+/// Wrapper type for `ty::TraitRef` to force printing the full path instead
+/// of a possibly unique short path.
+#[derive(Copy, Clone, TypeFoldable, Lift)]
+pub struct TraitRefPrintFullPath<'tcx>(ty::TraitRef<'tcx>);
+
+impl fmt::Debug for TraitRefPrintFullPath<'tcx> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Display::fmt(self, f)
+    }
+}
+
+impl ty::TraitRef<'tcx> {
+    pub fn print_full_path(self) -> TraitRefPrintFullPath<'tcx> {
+        TraitRefPrintFullPath(self)
     }
 }
 
 impl ty::Binder<ty::TraitRef<'tcx>> {
     pub fn print_only_trait_path(self) -> ty::Binder<TraitRefPrintOnlyTraitPath<'tcx>> {
         self.map_bound(|tr| tr.print_only_trait_path())
     }
+    pub fn print_only_trait_path_full(self) -> ty::Binder<TraitRefPrintOnlyTraitPath<'tcx>> {
+        self.map_bound(|tr| tr.print_only_trait_path_full())
+    }
+    pub fn print_full_path(self) -> ty::Binder<TraitRefPrintFullPath<'tcx>> {
+        self.map_bound(|tr| tr.print_full_path())
+    }
```
These changes shouldn't be necessary once we stop using string matching for diagnostics, AFAICT.

estebank replied:
We still need to have access to the full path for structured suggestions.

eddyb replied:
Presumably that can be handled by "just" using the "include `crate`" mode that these internally use, directly?

petrochenkov commented:
@da-x 
Could you assign this PR to me once it's otherwise ready?
Looks like things are continuing getting added to resolve and even crate loader.

da-x commented:
Self-reviewed again. 372 pages on a 100-line terminal :weight_lifting: 

1) Should rustfix handling be fixed (see `src/test/ui/bound-suggestions.fixed`) - Rustfix sees that `Debug` is needed, so it adds `Debug` without a `use std::fmt::Debug`, or alternatively refer to `std::fmt::Debug` directly.
2) Suggestions gone from `src/test/ui/range/issue-54505-no-literals.rs` (e.g. `SUGGESTION &std::ops::RangeTo { end: 5 }`).
3) Is it okay that it works for symbols such as `std::fn::drop_in_place` too?
4) Is more testing needed?

bors commented:
:umbrella: The latest upstream changes (presumably #74342) made this pull request unmergeable. Please resolve the merge conflicts.

da-x commented:
Thanks to @estebank, reverted trait_selection hacks, and we are down to `183 insertions(+), 17 deletions(-)` amount of non-test changes in the PR.

eddyb commented on src/librustc_typeck/check/op.rs:
```diff
@@ -562,7 +563,8 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                    on the left and may require reallocation. This \
                    requires ownership of the string on the left";
 
-        let is_std_string = |ty| &format!("{:?}", ty) == "std::string::String";
+        let is_std_string =
+            |ty| with_crate_prefix(|| &format!("{:?}", ty) == "std::string::String");
```
Should be diagnostic item.

da-x replied:
Indeed, like it was done for traits - I'll try to fix it in a separate PR.

da-x replied:
Pushed as #74471

eddyb commented on src/test/incremental/dirty_clean.rs:
```diff
@@ -27,14 +27,14 @@ mod y {
     pub fn y() {
-        //[cfail2]~^ ERROR `typeck_tables_of(y::y)` should be clean but is not
+        //[cfail2]~^ ERROR `typeck_tables_of(y)` should be clean but is not
         x::x();
     }
 }
 
 mod z {
     #[rustc_dirty(label="typeck_tables_of", cfg="cfail2")]
     pub fn z() {
-        //[cfail2]~^ ERROR `typeck_tables_of(z::z)` should be dirty but is not
+        //[cfail2]~^ ERROR `typeck_tables_of(z)` should be dirty but is not
```
These don't seem unambiguous, how does this happen? Maybe we should ignore namespaces, to be conservative?
(Although presumably this has been discussed already)

da-x replied:
So it looks like function name `z` and module name `z` are not colliding on the disambiguation check. If they were, then `z` name would have been excluded from path shortening. Not sure why yet. Functions are in `ValueNS` AFAIK, and modules - where? Maybe under `TypeNS` with `DefKind::Mod`? I'll take a closer look.

da-x replied:
This is fixed as of a few days ago.

bors commented:
:umbrella: The latest upstream changes (presumably #72983) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #74060) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #74471) made this pull request unmergeable. Please resolve the merge conflicts.

da-x commented:
@petrochenkov and others, I think only the main issue remaining is the effect over rustfix. I'd be glad for some suggestions, perhaps promoting `disable_unique_symbols` to user flags?

bors commented:
:umbrella: The latest upstream changes (presumably #72488) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #74785) made this pull request unmergeable. Please resolve the merge conflicts.

bors commented:
:umbrella: The latest upstream changes (presumably #74210) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented on src/librustc_resolve/diagnostics.rs:
```diff
@@ -870,7 +877,9 @@ impl<'a> Resolver<'a> {
                 }
-                if let Some(crate_id) = self.crate_loader.maybe_process_path_extern(ident.name) {
+                if let Some(crate_id) =
+                    self.crate_loader.maybe_process_path_extern(ident.name, self.avoid_crate_load)
```
This PR should not modify anything in `rustc_metadata`.

If `find_unique_symbols` runs during any execution and not only when errors are guaranteed to happen, then it should either
- not rely on `lookup_import_candidates` because it loops through the whole `self.extern_prelude`.
- loop only through crates that are already loaded, using `cstore.crates_untracked()` or other function listing the loaded `CrateNum`s.

da-x replied:
Based on this, found a way to modify `lookup_import_candidates` so that `rustc_metadata` can stay the same. Done in a subsequent commit.

bors commented:
:umbrella: The latest upstream changes (presumably #74948) made this pull request unmergeable. Please resolve the merge conflicts.

da-x commented:
r? @petrochenkov

Aaron1011 commented on src/librustc_session/options.rs:
```diff
@@ -935,6 +935,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "run LLVM in non-parallel mode (while keeping codegen-units and ThinLTO)"),
     no_profiler_runtime: bool = (false, parse_no_flag, [TRACKED],
         "prevent automatic injection of the profiler_builtins crate"),
+    disable_unique_symbols: bool = (false, parse_bool, [UNTRACKED],
```
Could this be changed to `enable_unique_symbols` or `use_unique_symbols`? I find it easier to deal with command-line arguments that don't involve double-negatives (e.g. "don't disable unique symbols").

da-x replied:
I understand the concern, but as far I can tell there's no double negative when passing `-Z disable-unique-symbols` to rustc in the command line. Is there another way to pass this?

Aaron1011 replied:
My concern is mainly when checking the flag - I think `!disable_unique_symbols` is slightly harder to understand than `enable_unique_symbols`. However, I don't have a strong opinion about it.

petrochenkov replied:
The option may make sense if you have all the context of this PR, but is really ambiguous otherwise.
"Symbol" in particular is a very overloaded term.

I suggest changing the user-visible option to `-Z trim-diagnostic-paths`.
This way we make more obvious what this option is about and don't promise the specific algorithm used for the trimming.

bors commented:
:umbrella: The latest upstream changes (presumably #74821) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
I went through the discussion thread and marked most of the review comments as resolved.
Remaining comments with unclear status:
- https://github.com/rust-lang/rust/pull/73996#discussion_r453490236

petrochenkov commented on src/test/mir-opt/unusual_item_types.core.ptr-drop_in_place.Vec_i32_.AddMovesForPackedDrops.before.mir.64bit:
```diff
@@ -0,0 +1,46 @@
+// MIR for `drop_in_place` before AddMovesForPackedDrops
```
Not sure why these two new files (`src/test/mir-opt/unusual_item_types.core.ptr-drop_in_place.Vec_i32_.AddMovesForPackedDrops.before.mir.32bit` and `src/test/mir-opt/unusual_item_types.core.ptr-drop_in_place.Vec_i32_.AddMovesForPackedDrops.before.mir.64bit`) appeared.
Perhaps it's a rebase artifact.

da-x replied:
Indeed - fixed in newly added merge.

petrochenkov commented on src/librustc_session/options.rs:
```diff
@@ -935,6 +935,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "run LLVM in non-parallel mode (while keeping codegen-units and ThinLTO)"),
     no_profiler_runtime: bool = (false, parse_no_flag, [TRACKED],
         "prevent automatic injection of the profiler_builtins crate"),
+    disable_unique_symbols: bool = (false, parse_bool, [UNTRACKED],
```
@da-x 
Have you seen the comments from the start and from the end of this list?
```
    // This list is in alphabetical order.
    //
    // If you add a new option, please update:
    // - src/librustc_interface/tests.rs
```
(The list of tests is also in alphabetical order.)

da-x replied:
Fixed in a new commit.

petrochenkov commented:
(I'll review the main part of the PR next weekend or during the week.)

da-x commented:
@Aaron1011, with the changes in this PR, `src/test/ui/proc-macro/meta-macro-hygiene.output` stopped being stable unless I pass `-Z disable-unique-symbols`. Got any idea?

Aaron1011 commented:
@da-x: What are the different outputs that you're getting?

da-x commented:
@Aaron1011 The list explodes with many entries - [see direct link to CI](https://pipelines.actions.githubusercontent.com/a51pe664LxB0T6W5TMkVTySSEHV6Mu7MfWbWvOoexbCw8EuBh3/_apis/pipelines/1/runs/10604/signedlogcontent/8?urlExpires=2020-08-13T09%3A05%3A34.2645782Z&urlSigningMethod=HMACV1&urlSignature=J%2FDvwBerRMUdwf79n%2FHFraJWbuFXIMEzdsUsWFqUeIY%3D). My guess is that the iteration done in `find_unique_symbols` adds to the list that goes into this output, collected by combining `-Z span-debug -Z macro-backtrace -Z unpretty=expanded,hygiene`, unique to this specific test. Previously, this code only ran if the compiler wanted to output errors.

bors commented:
:umbrella: The latest upstream changes (presumably #75322) made this pull request unmergeable. Please resolve the merge conflicts.

Aaron1011 commented:
@da-x: Yeah, it looks like `lookup_import_candidates_from_module` calls `for_each_child` on the `std` crate, which results in us loading hygiene for a bunch of exports (which I was trying to avoid by using `#![no_std]`).

I think it's perfectly fine to add `-Z disable-unique-symbols` to that test, since that test is really only about macro output and explicitly loaded hygiene information.

estebank commented on src/librustc_resolve/diagnostics.rs:
```diff
@@ -725,17 +726,19 @@ impl<'a> Resolver<'a> {
         parent_scope: &ParentScope<'a>,
         start_module: Module<'a>,
         crate_name: Ident,
         filter_fn: FilterFn,
+        mut collect_fn: CollectFn,
```
I think that we could use `impl Trait` here, but it is a stylistic choice that we might or might not want.

```suggestion
        filter_fn: Fn(Res) -> bool,
        mut collect_fn: for<'b> FnMut(&'b Ident, Namespace, &'b NameBinding<'a>),
```

da-x replied:
It's very helpful. Done.

estebank commented on src/librustc_resolve/lib.rs:
```diff
@@ -917,11 +917,16 @@ pub struct Resolver<'a> {
 
+    /// Names that uniquely identify an importable definition.
+    unique_symbols: FxHashMap<DefId, Symbol>,
+    avoid_crate_load: bool,
```
Can we document this field?

da-x replied:
Done

estebank commented:
At some point we'll need to squash the commits and rebase them on top of master? The non-diagnostic related changes seem tightly related so they feel to me to belong in the same commit (although @petrochenkov is the reviewer so follow what they ask for).

The code looks reasonable to me.

@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit b1c323ba83c5690af6d45ed2f2fa693ab5bf519c with merge e6395ff11c83aa726d50a432bd95d0a1dbecc20c...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b1c323ba83c5690af6d45ed2f2fa693ab5bf519c","merge_sha":"e6395ff11c83aa726d50a432bd95d0a1dbecc20c"} -->

estebank commented:
```
tidy check
* 615 error codes
* highest error code: E0772
Checking which error codes lack tests...
Found 509 error codes
Found 0 error codes with no tests
Done!
tidy error: /checkout/src/test/ui/proc-macro/meta-macro-hygiene.rs:4: line longer than 100 chars
some tidy checks failed
```

Also, @bors rollup=never

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/986206331), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35993)
Build commit: e6395ff11c83aa726d50a432bd95d0a1dbecc20c (`e6395ff11c83aa726d50a432bd95d0a1dbecc20c`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=35993","checks-actions":"https://github.com/rust-lang-ci/rust/runs/986206331"},"merge_sha":"e6395ff11c83aa726d50a432bd95d0a1dbecc20c"} -->

rust-timer commented:
Queued e6395ff11c83aa726d50a432bd95d0a1dbecc20c with parent 55b9adfafa11b2ced5c0477c949fd875b19b3877, future [comparison URL](https://perf.rust-lang.org/compare.html?start=55b9adfafa11b2ced5c0477c949fd875b19b3877&end=e6395ff11c83aa726d50a432bd95d0a1dbecc20c).

rust-timer commented:
Finished benchmarking try commit (e6395ff11c83aa726d50a432bd95d0a1dbecc20c): [comparison url](https://perf.rust-lang.org/compare.html?start=55b9adfafa11b2ced5c0477c949fd875b19b3877&end=e6395ff11c83aa726d50a432bd95d0a1dbecc20c).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

petrochenkov commented:
@da-x 
Meta: could you avoid merge commits? (Rebase the changes on master and add the new commits on top of that.)
In the end, before the merge, the commits will also need to be stashed.

da-x commented:
@petrochenkov No problem, I can work with rebasing. This PR has high conflict potential due to tests so it can help to avoid all of these accumulating merge commits.

Note that the squash was forced-push, along with the following:

* I've put a big comment above `find_unique_symbols` with the knowledge and changes accumulated so far, describing what it does.
* Fixes per @estebank's review.

petrochenkov commented on src/librustc_resolve/diagnostics.rs:
```diff
@@ -857,54 +858,173 @@ impl<'a> Resolver<'a> {
+    ///   `std::fmt::Display` and `std::path::Display`, for instance.
+    /// * Skip Variant types. Othersize, it can pick up `core::num::flt2dev::Part::Copy` as collsion
+    ///   with `std::marker::Copy`. It also makes sense, as users cannot specifically import this
+    ///   into the Type namespace (rather than the Value namespace).
```
Looks like you want to special case some well known names (i.e. `Copy` is always trimmed), but do that with more general heuristics that could be wrong for some other user-defined items that are not well-known.

I would suggest to get rid of all these heuristics in the initial PR (and remove `enum Kind`).
If two items with the same name have different `DefIds`, the I suggest to consider them non-unique.
The heuristics can be submitted as separate PRs later, then their impact could be reviewed better.

da-x replied:
Done - separated away from this PR, and prepared another commit for later.

da-x replied:
(waiting in a side-branch: https://github.com/rust-lang/rust/commit/fb92566fd53f89dc158096e242bc5578a3971d42)

petrochenkov commented on src/librustc_resolve/diagnostics.rs:
```diff
@@ -857,54 +858,173 @@ impl<'a> Resolver<'a> {
+            Ident::invalid(),
+            TypeNS,
+            &parent_scope,
+            |_| false,
```
I'm not sure reusing `lookup_import_candidates` is a good idea here.
All its arguments end up being unused, basically, except for `parent_scope` which ends up being used incorrectly.

Could you keep `lookup_import_candidates` alone, and introduce a new function that visits items in all crates like `lookup_import_candidates`, but drops all the activities not necessary for `find_unique_symbols`?
(We may need to do that for performance anyway.)

petrochenkov replied:
(Then `avoid_crate_load` will become entirely unnecessary as well.)

da-x replied:
Done. The new function's implementation is shorter and may be faster.

petrochenkov commented:
The performance (https://github.com/rust-lang/rust/pull/73996#issuecomment-674255345) is pretty disastrous.
Perhaps things will become better with the suggestion from https://github.com/rust-lang/rust/pull/73996#discussion_r471035977 implemented.

da-x commented on src/librustc_middle/ty/print/pretty.rs:
```diff
@@ -243,6 +243,20 @@ pub trait PrettyPrinter<'tcx>:
+    fn try_print_unique_def(mut self, def_id: DefId) -> Result<(Self::Path, bool), Self::Error> {
+        let unique_symbols = &self.tcx().unique_symbols;
+
+        match unique_symbols.get(&def_id) {
```
Done

da-x commented on src/librustc_middle/ty/print/pretty.rs:
```diff
@@ -243,6 +243,20 @@ pub trait PrettyPrinter<'tcx>:
 
+    /// Try to see if this is a unique symbol, meaning there's no other definition with
+    /// the same name.
+    fn try_print_unique_def(mut self, def_id: DefId) -> Result<(Self::Path, bool), Self::Error> {
```
Done

da-x commented on src/librustc_resolve/diagnostics.rs:
```diff
@@ -857,54 +858,173 @@ impl<'a> Resolver<'a> {
+        let unique_symbols: &mut FxHashMap<(Kind, Symbol), Option<DefId>> =
+            &mut FxHashMap::default();
+
+        for symbol in self.glob_imported.drain() {
```
Done

da-x commented:
@petrochenkov Done changes requested. In case that the performance impact is still problematic, I guess that storing unique symbols in each crate's meta-data would help to make it faster?

petrochenkov commented:
From the new implementation you can see how little resolver is actually used from `fn find_unique_symbols`.
It's very likely that the same logic can be implemented without resolver's help at all, with only `tcx` available.

In that case the most reliable way to fix the performance issue would be to build the `unique_symbols` table on demand, when the first def path is printed, and keep it somewhere in `tcx`.
(That's unless def paths are pretty-printed during normal compilation, which would be strange, but who knows.)

estebank commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion

bors commented:
:hourglass: Trying commit 93510ad3841cfed04e3a0e5d4d9aab9b127e1e87 with merge 0f8576a22c16ed36c3a48a0e08babed4ea521302...
<!-- homu: {"type":"TryBuildStarted","head_sha":"93510ad3841cfed04e3a0e5d4d9aab9b127e1e87","merge_sha":"0f8576a22c16ed36c3a48a0e08babed4ea521302"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/991496504), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36033)
Build commit: 0f8576a22c16ed36c3a48a0e08babed4ea521302 (`0f8576a22c16ed36c3a48a0e08babed4ea521302`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36033","checks-actions":"https://github.com/rust-lang-ci/rust/runs/991496504"},"merge_sha":"0f8576a22c16ed36c3a48a0e08babed4ea521302"} -->

rust-timer commented:
Queued 0f8576a22c16ed36c3a48a0e08babed4ea521302 with parent 9b4db695b0ab13885a61deb1b2e4d6599b8c5bbc, future [comparison URL](https://perf.rust-lang.org/compare.html?start=9b4db695b0ab13885a61deb1b2e4d6599b8c5bbc&end=0f8576a22c16ed36c3a48a0e08babed4ea521302).

rust-timer commented:
Finished benchmarking try commit (0f8576a22c16ed36c3a48a0e08babed4ea521302): [comparison url](https://perf.rust-lang.org/compare.html?start=9b4db695b0ab13885a61deb1b2e4d6599b8c5bbc&end=0f8576a22c16ed36c3a48a0e08babed4ea521302).

Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. Please note that if the perf results are neutral, you should likely undo the rollup=never given below by specifying `rollup-` to bors.

Importantly, though, if the results of this run are non-neutral **do not** roll this PR up -- it will mask other regressions or improvements in the roll up.

@bors rollup=never

da-x commented:
Re-implement using `tcx`, as a new on-demand query.

petrochenkov commented:
>(That's unless def paths are pretty-printed during normal compilation, which would be strange, but who knows.)

@da-x 
I was terribly wrong, path pretty-printing is used all the time :(
It's probably not called *often*, but it is, for example, called for almost all queries in `src\librustc_middle\query\mod.rs` to build query descriptions.
This means on-demand won't help, the unique symbol collection will always run anyway.

So, I guess storing the unique symbol information in metadata as suggested in https://github.com/rust-lang/rust/pull/73996#issuecomment-674569533 is the only remaining way to make this cheaper.

da-x commented:
> > (That's unless def paths are pretty-printed during normal compilation, which would be strange, but who knows.)
> 
> @da-x
> I was terribly wrong, path pretty-printing is used all the time :(
> It's probably not called _often_, but it is, for example, called for almost all queries in `src\librustc_middle\query\mod.rs` to build query descriptions.

If it's called all the time for internal reasons, it is odd to me that the changes in this PR would affect the output.

Maybe like `with_crate_prefix` we should have `with_user_facing` that would explicitly allow `try_print_trimmed_def_path` to take place, so it can be turned on, only when generating error messages, so then it would become on-demand only in that case? Alternatively, instead of wrapper, we can use a global flag in `tcx` that is turned on just before beginning to emit messages, after all the internal queries are done. 

If you're sure it's not salvageable, I'll try the meta-data tweaks instead.

bors commented:
:umbrella: The latest upstream changes (presumably #74748) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented:
@da-x 
>Maybe like `with_crate_prefix` we should have `with_user_facing` that would explicitly allow `try_print_trimmed_def_path` to take place, so it can be turned on, only when generating error messages, so then it would become on-demand only in that case? 

You could turn `try_print_trimmed_def_path` into a panic and try to build at least libcore to estimate the amount of work required to this.

The on-demand scheme is also a bit fragile if the sense that any careless use of `print_def_path` will introduce a big perf regression.

One variation on this theme is to always use `tcx.trimmed_def_path_str(...)` explicitly and manually.

>Alternatively, instead of wrapper, we can use a global flag in `tcx` that is turned on just before beginning to emit messages, after all the internal queries are done. 

This is an interesting suggestion - turn on the trimming once any error is reported.

One issue here though is that the def path string is usually constructed *right before* emitting an error showing that string, so the first errors won't generally be trimmed (unless all the affected error reporting code is restructured to produce def path strings later).

da-x commented:
After some investigation, I have found there is a lot of reliance over `Display` trait to convert various types to strings, and some of these conversions contain paths. Some of the conversions are internal for debugging, and others also are emitted by diagnostics. The call sites divide to various classes:

- `format!()`
- `.is_string()`
- Explicit calls to `def_path_str` and its variants.

Some are done way before constructions of `DiagnosticsBuilder`, so simply turning on a global flag in `DiagnosticsBuilder::new` is insufficient. Some `DiagnosticsBuilder` are lazy in a sense that they are 'hidden' and kept aside while no `.emit()` takes place.

@petrochenkov are you in favor of the idea to explicitly change these hundred of calls sites to new explicit APIs? I.e,

- 'user_format!()`, relaying to format!() and affecting its context via TLS.
- `.is_user_string()`, using new `IsUserString` trait, affecting its context via TLS.
- Having `user_path_str` variants.

Also, to prevent a performance regression, add a check at the end of the build to see that the `unique_symbols` map was constructed if and only if diagnostics were printed.

Alternatively we can have `user_format_context!()` macro added in the beginning of various query functions in diagnostics using RAII trick to affect the Display trait via TLS. It would mean a lesser degree of code changes. However, `debug!` and friends would emit the trimmed path, so perhaps the explicit approach is more desired.

In parallel, I'm also checking the metadata approach (though it is not clear that hashset operations over dependent crates the `find_unique_symbols` won't be a performance issue too).

petrochenkov commented:
>are you in favor of the idea to explicitly change these hundred of calls sites to new explicit APIs?

No, I'm certainly not in favor of changing hundreds of calls.
Doing this transparently would be the best outcome, changing a few places would be ok, rewriting all the diagnostics is something to avoid unless there's not other way.

da-x committed:
rustc_{errors,session}: add `delay_good_path_bug`

The first use case of this detection of regression for trimmed paths
computation, that is in the case of rustc, which should be computed only
in case of errors or warnings.

Our current user of this method is deeply nested, being a side effect
from `Display` formatting on lots of rustc types. So taking only the
caller to the error message is not enough - we should collect the
traceback instead.

da-x commented:
Despite CI success, with `debug-assertions = true` I see two failures in `incremental` tests, the simple of them is the following:

```
stderr:
------------------------------------------
error: CGU-reuse for `krate_inherent-x` is `No` but should be at least `PreLto`
  --> /home/dan/dev/rust/rust/src/test/incremental/krate-inherent.rs:7:1
   |
LL | #![rustc_partition_reused(module = "krate_inherent-x", cfg = "cfail2")]
```

Not sure yet how to debug this. Maybe @nikomatsakis has some pointers on where I should look?

petrochenkov commented:
I see you've special cased rustdoc with regards to these trimmed path, but what about other tools? They cannot all be special cased.

I see at least three scenarios for tool preferences:
- Pretty-printed paths are not a core part of the tool's functionality (example: rustc). Then it can:
    - Disable trimming and be fast.
    - Enable trimming and be slow.
    - Enable trimming carefully, use exceptions and `delay_good_path_bug`.
- Pretty-printing paths is a core part of the tool's functionality. Then it can:
    - Disable trimming and be fast.
    - Enable trimming and be slow.

This behavior should probably be configured through `rustc_driver` somehow.

estebank commented:
> I only have concerns about considering `save_analysis` a mode that can be slowed down, otherwise it all seems reasonable.

We could make that mode that mode effectively imply the full path flag being introduced, right?

petrochenkov commented:
@estebank 
>We could make that mode that mode effectively imply the full path flag being introduced, right?

Yes, for each mode there's always a choice where it can go - into `try_print_trimmed_def_path` or into `delay_good_path_bug`.
(Although it could make sense to just wrap the whole `save::process_crate` into `with_no_trimmed_path`).

bors commented:
:umbrella: The latest upstream changes (presumably #75302) made this pull request unmergeable. Please resolve the merge conflicts.

da-x committed:
Fix some unwanted uses of Debug formatting on user-facing messages

While formatting for user diagnostics used `Display` for all most cases,
some small amount of cases used `Debug` instead.  Until now, `Display`
and `Debug` yielded the same output for many types. However, with path
trimming, we want to show a shorter path for the user, these cases need
fixing.

da-x commented:
* Rebased as needed.
* Took suggestion for `with_no_trimmed_path` on `save::process_crate`, so no need to refer to `save_analysis` from `delay_good_path_bug`.
* `incremental` tests are working now, after I've added path trimming exceptions for `Debug` impls of types in `structural_impls.rs`. It should be made in a different PR to find how come the tests were sensitive to `Debug` impls in the first place, and why only with `debug-assertions`.
* This reveals a few cases where `Debug` output went into standard diagnostics messages of rustc users, so these are fixed in a separate commit. It may be a few of the 19 `with_no_trimmed_path` calls turned redundant - I have not verified.
* Instead of special casing `rustdoc`, added another boolean, which only `rustc`, our compilation critical path, turns off. I'm not sure whether we should keep `disable-unique-symbols` as debug, or maybe change to be a non-debug flag and have some tools use it.

bors commented:
:umbrella: The latest upstream changes (presumably #74862) made this pull request unmergeable. Please resolve the merge conflicts.

petrochenkov commented on src/librustc_session/config.rs:
```diff
@@ -606,6 +606,7 @@ impl Default for Options {
             unstable_features: UnstableFeatures::Disallow,
             debug_assertions: true,
             actually_rustdoc: false,
+            allow_good_path_bugs: true,
```
So with these defaults tools use path trimming and are slow by default.
I think it would be a safer default for tools to not use path trimming at all.

So the option needs to be about path printing specifically rather than about good path bugs in general, and it also needs to be a three variant enum rather than a boolean (https://github.com/rust-lang/rust/pull/73996#issuecomment-679355482).
```rust
enum TrimmedDefPaths {
    Never, // [default] `try_print_trimmed_def_path` never prints a trimmed path and never calls the expensive query
    Always, // `try_print_trimmed_def_path` calls the expensive query, the query doesn't call `delay_good_path_bug`
    GoodPath, // `try_print_trimmed_def_path` calls the expensive query, the query calls `delay_good_path_bug`
}
```

petrochenkov commented on src/librustc_session/options.rs:
```diff
@@ -825,6 +828,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
     dep_tasks: bool = (false, parse_bool, [UNTRACKED],
         "print tasks that execute and the color their dep node gets (requires debug build) \
         (default: no)"),
+    disable_unique_symbols: bool = (false, parse_bool, [UNTRACKED],
```
```suggestion
    trim_diagnostic_paths: bool = (true, parse_bool, [UNTRACKED],
```
https://github.com/rust-lang/rust/pull/73996#discussion_r471036398

da-x commented:
@petrochenkov My push today addressed the review comments. Remaining is perhaps to consider naming consistency of `trimmed_def_paths` vs. `with_no_trimmed_path` vs. `trim_diagnostic_paths`. I don't hold a strong opinion about this.

petrochenkov commented on compiler/rustc_interface/src/tests.rs:
```diff
@@ -468,6 +468,7 @@ fn test_debugging_options_tracking_hash() {
     untracked!(borrowck_stats, true);
     untracked!(deduplicate_diagnostics, true);
     untracked!(dep_tasks, true);
+    untracked!(trim_diagnostic_paths, false);
```
The option list needs sorting.

petrochenkov commented on compiler/rustc_session/src/options.rs:
```diff
@@ -825,6 +828,8 @@ options! {DebuggingOptions, DebuggingSetter, basic_debugging_options,
         "print tasks that execute and the color their dep node gets (requires debug build) \
         (default: no)"),
+    trim_diagnostic_paths: bool = (true, parse_bool, [UNTRACKED],
+        "in diagnostics, enable shortening of paths to unique symbols"),
```
The option list needs sorting.

petrochenkov replied:
```suggestion
        "in diagnostics, use heuristics to shorten paths referring to items"),
```
Or something like that.
"Unique symbols" are an implementation detail.

petrochenkov commented:
Ok, it's time to finally land this!
@bors r+

(I'll re-approve this if you fix the remaining nits.)

bors commented:
:pushpin: Commit e156318660590f1ef86aeaeb57d8dbf957012ee2 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov e156318660590f1ef86aeaeb57d8dbf957012ee2 -->
<!-- homu: {"type":"Approved","sha":"e156318660590f1ef86aeaeb57d8dbf957012ee2","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit e156318660590f1ef86aeaeb57d8dbf957012ee2 with merge 5eab51c11f41d7ccde88ea5c6c53a92cc8328037...
<!-- homu: {"type":"BuildStarted","head_sha":"e156318660590f1ef86aeaeb57d8dbf957012ee2","merge_sha":"5eab51c11f41d7ccde88ea5c6c53a92cc8328037"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1058783053)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1058783053","builder_name":"checks-actions"} -->

aDotInTheVoid commented:
Build is complaining about the broken link [here](https://github.com/da-x/rust/blob/e156318660590f1ef86aeaeb57d8dbf957012ee2/compiler/rustc_session/src/config.rs#L319)
<details>

```
 error: unresolved link to `default`
   --> compiler/rustc_session/src/config.rs:319:10
    |
319 |     /// [default] `try_print_trimmed_def_path` never prints a trimmed path and never calls the expensive query
    |          ^^^^^^^ unresolved link
    |
    = note: `-D broken-intra-doc-links` implied by `-D warnings`
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\
```

</details>

eddyb commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -892,6 +910,16 @@ impl HandlerInner {
         self.delay_as_bug(diagnostic)
     }
 
+    #[track_caller]
```
The `#[track_caller]` on `delay_span_bug` is used only for  `std::panic::Location::caller()`, which is here replaced with a backtrace.

eddyb commented on compiler/rustc_errors/src/lib.rs:
```diff
@@ -892,6 +910,16 @@ impl HandlerInner {
+        if self.flags.report_delayed_bugs {
+            self.emit_diagnostic(&diagnostic);
+        }
+        diagnostic.note(&format!("delayed at {}", std::backtrace::Backtrace::force_capture()));
```
This might need to account for formatting, especially around newlines. But also, is this necessary? If you can force the above `emit_diagnostic` to run, wouldn't that give you a backtrace (and a query stack, etc.)? Or is this for diagnosing issues on CI?

da-x replied:
The purpose of the backtrace here is to capture the full call site, but to not print anything, because it is not yet an error, so forcing the call to `emit_diagnostic` would not serve this purpose. It popped in CI when I was working on the PR and the formatting worked out alright.

eddyb commented:
I was confused by "good path", and looking a bit more, I'm not sure whether "path" refers to a Rust path as in `a::b::c` or whether it's "happy path" (which, IIRC, is what I usually see it as, not "good path" - there might also be something involving "success").

As `delay_good_path_bug` is only called once, couldn't you inline the conditions and use `delay_span_bug` directly? That would remove all occurrences of `good_path`, and we're left with `TrimmedDefPaths::GoodPath` which...

Oh, looking at this closer, I just realized a bigger problem than all of those special-cased conditions at the start of `delay_good_path_bug`: warnings (which @da-x did mention, now that I know to look for it).

In that case, even "happy path" or "success path" don't seem to encompass the correct situation.
So I'd call it `delay_bug_if_no_user_facing_output` or something similarly verbose, for clarity.

da-x commented:
@eddyb Most concise alternative I can think of is `messageless_run`, perhaps

eddyb commented:
@da-x This is a rare enough name (and completely internal, AFAICT) that a very verbose name wouldn't hurt anything.

da-x committed:
pretty: trim paths of unique symbols

If a symbol name can only be imported from one place for a type, and
as long as it was not glob-imported anywhere in the current crate, we
can trim its printed path and print only the name.

This has wide implications on error messages with types, for example,
shortening `std::vec::Vec` to just `Vec`, as long as there is no other
`Vec` importable anywhere.

This adds a new '-Z trim-diagnostic-paths=false' option to control this
feature.

On the good path, with no diagnosis printed, we should try to avoid
issuing this query, so we need to prevent trimmed_def_paths query on
several cases.

This change also relies on a previous commit that differentiates
between `Debug` and `Display` on various rustc types, where the latter
is trimmed and presented to the user and the former is not.

da-x commented:
@petrochenkov, @eddyb I've applied suggested fixes, and extended docstrings for new internal APIs `with_no_trimmed_paths` and `Session::delay_good_path_bug`.

petrochenkov commented:
@bors r+

Some internal renaming suggested in https://github.com/rust-lang/rust/pull/73996#issuecomment-685273159 still wasn't applied.
Remaining unresolved comment - https://github.com/rust-lang/rust/pull/73996#discussion_r481618671.

bors commented:
:pushpin: Commit 5203a6a650b349c3dcd876ee81a5e09c76791f93 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 5203a6a650b349c3dcd876ee81a5e09c76791f93 -->
<!-- homu: {"type":"Approved","sha":"5203a6a650b349c3dcd876ee81a5e09c76791f93","approver":"petrochenkov"} -->

da-x commented:
@eddyb not sure exactly how to name it, but I have some dislike for very long names even for a single use. There will be a follow up PR anyway for this, for fine tuning the unique symbols search. We can decide until this?

bors commented:
:hourglass: Testing commit 5203a6a650b349c3dcd876ee81a5e09c76791f93 with merge 2db2c87d3c1867b7cbd4578efcbf555454a36cce...
<!-- homu: {"type":"BuildStarted","head_sha":"5203a6a650b349c3dcd876ee81a5e09c76791f93","merge_sha":"2db2c87d3c1867b7cbd4578efcbf555454a36cce"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1062298809)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1062298809","builder_name":"checks-actions"} -->

petrochenkov commented:
`x.py test src/test/ui --compare-mode nll` tests need an update.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 07e7823c01be1733df2480de19fbbe6b8e9384cf has been approved by `petrochenkov`

<!-- @bors r=petrochenkov 07e7823c01be1733df2480de19fbbe6b8e9384cf -->
<!-- homu: {"type":"Approved","sha":"07e7823c01be1733df2480de19fbbe6b8e9384cf","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit 07e7823c01be1733df2480de19fbbe6b8e9384cf with merge 9698ddfa7daaaf2eb6a60d35aac2b7ece7d42d66...
<!-- homu: {"type":"BuildStarted","head_sha":"07e7823c01be1733df2480de19fbbe6b8e9384cf","merge_sha":"9698ddfa7daaaf2eb6a60d35aac2b7ece7d42d66"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1064028850)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1064028850","builder_name":"checks-actions"} -->

da-x commented:
We are at tools now. Rebuilding locally with `extended = true`.

da-x committed:
rustc_driver: have TrimmedDefPaths::GoodPath only for rustc

`run_compiler` is used by clippy and other tools, which should not have
the trimmed paths feature enabled by default, until we see it works well
for them.

Would also be nice to rename `TimePassesCallbacks` however it's a
submodule change.

petrochenkov commented:
Wait a second, how can tools fail if `TrimmedDefPaths::Never` is the default for everything except for rustc.

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit c5a61319da34e931072f2ffe3d12e530034c8c07 has been approved by `petrochenkov`

<!-- @bors r=petrochenkov c5a61319da34e931072f2ffe3d12e530034c8c07 -->
<!-- homu: {"type":"Approved","sha":"c5a61319da34e931072f2ffe3d12e530034c8c07","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit c5a61319da34e931072f2ffe3d12e530034c8c07 with merge eb675917e4b586989f41f4615494dbb24fff6a23...
<!-- homu: {"type":"BuildStarted","head_sha":"c5a61319da34e931072f2ffe3d12e530034c8c07","merge_sha":"eb675917e4b586989f41f4615494dbb24fff6a23"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1066036952)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1066036952","builder_name":"checks-actions"} -->

kennytm commented:
@bors retry

```
2020-09-03T09:22:06.0783173Z [0m[0m[1m[32m Downloading[0m crates ...
2020-09-03T09:22:06.2441826Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (2 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/byteorder/1.3.4/download`, got 503
2020-09-03T09:22:06.2738086Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (2 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/typenum/1.12.0/download`, got 503
2020-09-03T09:22:06.2759970Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (2 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/either/1.6.0/download`, got 503
2020-09-03T09:22:06.2797654Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (2 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/opaque-debug/0.2.3/download`, got 503
2020-09-03T09:22:06.2899875Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (1 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/byteorder/1.3.4/download`, got 503
2020-09-03T09:22:06.2902921Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (1 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/typenum/1.12.0/download`, got 503
2020-09-03T09:22:06.2909162Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (1 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/either/1.6.0/download`, got 503
2020-09-03T09:22:06.2920125Z [0m[0m[1m[33mwarning[0m[1m:[0m spurious network error (1 tries remaining): failed to get 200 response from `https://crates.io/api/v1/crates/opaque-debug/0.2.3/download`, got 503
2020-09-03T09:22:06.4454797Z [0m[0m[1m[31merror[0m[1m:[0m failed to download from `https://crates.io/api/v1/crates/byteorder/1.3.4/download`
```

bors commented:
:hourglass: Testing commit c5a61319da34e931072f2ffe3d12e530034c8c07 with merge ecb4db95aeb8f9c94e2268d63ca39fdd70384477...
<!-- homu: {"type":"BuildStarted","head_sha":"c5a61319da34e931072f2ffe3d12e530034c8c07","merge_sha":"ecb4db95aeb8f9c94e2268d63ca39fdd70384477"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1066292665)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1066292665","builder_name":"checks-actions"} -->

da-x committed:
specialization_graph: avoid trimmed paths for OverlapError

da-x committed:
rustc_lint: avoid trimmed paths for ty_find_init_error

da-x commented:
`cargotest` building two crates found two more cases that formatted types to string without emitting errors. I fixed two but I didn't check the largest of the them - servo. I wonder what more cases would be revealed when trying to built `crates.io` :).

Luckily it only represents very uncommonly emitted diagnostics, so we can postpone handling these to subsequent PRs.

da-x commented:
As for the second, it's possible to put the whole of `late_lint_mod` under `with_no_trimmed_paths` but it looks too big of a hammer, affecting a few tens of tests.

bors commented:
:pushpin: Commit e7d7615105332f269a27cbd7273029377a96ccdf has been approved by `petrochenkov`

<!-- @bors r=petrochenkov e7d7615105332f269a27cbd7273029377a96ccdf -->
<!-- homu: {"type":"Approved","sha":"e7d7615105332f269a27cbd7273029377a96ccdf","approver":"petrochenkov"} -->

bors commented:
:hourglass: Testing commit e7d7615105332f269a27cbd7273029377a96ccdf with merge d10ea35cf6c949c06cf0a7856b5cad133c3feb43...
<!-- homu: {"type":"BuildStarted","head_sha":"e7d7615105332f269a27cbd7273029377a96ccdf","merge_sha":"d10ea35cf6c949c06cf0a7856b5cad133c3feb43"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1068521979)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1068521979","builder_name":"checks-actions"} -->

da-x commented:
Recent failure is a single test under `rustdoc-ui` that only fails on `i686-mingw-1`, and could be unrelated.

petrochenkov commented:
@bors r+
@bors delegate+

If that test fails on `i686-mingw-1` again you can ignore it and make an issue for fixing it later.

bors commented:
:bulb: This pull request was already approved, no need to approve it again.

- This pull request previously failed. You should add more commits to fix the bug, or use `retry` to trigger a build again.
- There's another pull request that is currently being tested, blocking this pull request: #73819

bors commented:
:pushpin: Commit e7d7615105332f269a27cbd7273029377a96ccdf has been approved by `petrochenkov`

<!-- @bors r=petrochenkov e7d7615105332f269a27cbd7273029377a96ccdf -->
<!-- homu: {"type":"Approved","sha":"e7d7615105332f269a27cbd7273029377a96ccdf","approver":"petrochenkov"} -->

bors commented:
:v: @da-x can now approve this pull request
<!-- homu: {"type":"Delegated","delegator":"petrochenkov","delegate":"da-x"} -->

bors commented:
:hourglass: Testing commit e7d7615105332f269a27cbd7273029377a96ccdf with merge af3c6e733a40e671550e0f0f5aeecaa13772ba56...
<!-- homu: {"type":"BuildStarted","head_sha":"e7d7615105332f269a27cbd7273029377a96ccdf","merge_sha":"af3c6e733a40e671550e0f0f5aeecaa13772ba56"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1069631723), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36398)
Approved by: petrochenkov
Pushing af3c6e733a40e671550e0f0f5aeecaa13772ba56 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36398","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1069631723"},"merge_sha":"af3c6e733a40e671550e0f0f5aeecaa13772ba56"} -->

lukaslueg commented:
@da-x Op here, thanks for this :-)

ecstatic-morse commented:
Final perf results [are here](https://perf.rust-lang.org/compare.html?start=0d0f6b113047b2cf9afbde990cee30fd5b866469&end=af3c6e733a40e671550e0f0f5aeecaa13772ba56&stat=instructions:u). Seems like a small regression to instruction counts, but most of it is in incremental benchmarks, which run quite fast. Thanks @da-x!

tronta commented:
One question about the following error message:
```
error[E0609]: no field `fitnes` on type `&mut Sheep`
  --> src/main.rs:23:14
   |
23 |         self.fitnes += 1;
   |              ^^^^^^ help: a field with a similar name exists: `fitness`
```
Could this be also shorted to one 'Sheep' like
```
error[E0609]: no field `fitnes` on type `Sheep`
  --> src/main.rs:23:14
   |
23 |         self.fitnes += 1;
   |              ^^^^^^ help: a field with a similar name exists: `fitness`
```

lzutao commented:
This is an orthogonal issue. You could open an issue/PR to discuss.

bjorn3 commented:
This breaks debug printing of types without passing `-Ztrim-diagnostic-paths=no`.

bjorn3 commented on compiler/rustc_driver/src/lib.rs:
```diff
@@ -159,10 +160,7 @@ pub fn run_compiler(
-        trimmed_def_paths: TrimmedDefPaths::GoodPath,
-        ..config::build_session_options(&matches)
-    };
+    let sopts = config::build_session_options(&matches);
```
Why should tools not have trimmed paths enabled?
